src/v/config/configuration.cc

```diff
@@ -1282,6 +1282,53 @@ configuration::configuration()
        .example = "false",
        .visibility = visibility::user},
       true)
+  , auditing_enabled(
+      *this,
+      "auditing_enabled",
+      "Enables the audit logging subsystem",
+      {.needs_restart = needs_restart::no, .visibility = visibility::user},
+      false)
+  , audit_log_num_partitions(
+      *this,
+      "audit_log_num_partitions",
+      "Number of partitions for the internal audit log topic",
+      {.needs_restart = needs_restart::no, .visibility = visibility::user},
+      12)
+  , audit_log_replication_factor(
+      *this,
+      "audit_log_replication_factor",
+      "Replication factor of the internal audit log topic",
+      {.needs_restart = needs_restart::no, .visibility = visibility::user},
+      3)
+  , audit_client_max_buffer_size(
+      *this,
+      "audit_client_max_buffer_size",
+      "Maximum number of bytes the internal audit client will allocate for "
+      "audit log records",
+      {.needs_restart = needs_restart::no, .visibility = visibility::user},
+      1_MiB)
+  , audit_queue_drain_interval_ms(
+      *this,
+      "audit_queue_drain_interval_ms",
+      "Frequency in which per shard audit logs are batched to client for write "
+      "to audit log",
+      {.needs_restart = needs_restart::no, .visibility = visibility::tunable},
+      500ms)
+  , audit_max_queue_elements_per_shard(
+      *this,
+      "audit_max_queue_elements_per_shard",
+      "Max number of elements in the audit buffers before broker blocks on "
+      "registered auth behavior",
```

Comment:
Maybe a question for @deniscoady @michael-redpanda

Reply:
To answer the question though , yeah blocking means the request handler that is registered for auditing will stop accepting requests if buffers are full.

Reply:
I don't think there is probably a question about what these things are, it's a matter of writing out a description that is consumable beyond the core team as an audience. And maybe that means eliminating internal implementation details.

Reply:
Your asking
> I wonder if this particular option should be off by default. It seems like it is going to be hard to tune.

which to me implies is there a mode we can run in which wouldn't block if we cannot enqueue an audit message. Maybe i'm misunderstanding but it seems like this is a question for product because in the PRD this is a stated requirement.

Reply:
Good point. I guess it seems like "memory usage" rather than "number of elements" is the resource accounting that we'd want?

Reply:
And if we had both, I'd question the value of counting the number of elements as a packpressure signal if memory limits exist already.

Reply:
Yes exactly, I actually wanted to bring this up earlier. Currently there is a config option for number of elements but that doesn't really make sense, IMO we should do the estimation thing again with audit elements per shard. 

I got some pushback from @michael-redpanda about this though i forgot why we decided to go with number of elements instead of doing an estimate here as well.

Reply:
So there's two type of containers here:
* a buffer on each shard containing audit event metadata (some structurzed thing of data that we keep structurized so we can find repeated events and increment their count)
  * Size determined by number of elements 
  * The container is probably a boost::multi_index thing
* The client buffer on shard0 that contains the serialized data (effectively the data that goes on the wire)
  * Size determined by the size of the JSON serialized strings

I do recognize it's weird.  One buffer is count based and the other is size based.  And the back pressure is sourced from the size based one.

If we want to limit the size of the first buffer by it's _future_ memory usage (when the structures are serialized into JSON), we'd have to figure out how to estimate the size of the JSON.  One way would be to simply serialize it and record its size (wasteful cpu cycles?).  But that's not guaranteed to always be the size (e.g. if count >1, we have three more fields being populated in the final buffer).

Reply:
I don't think we should measure future memory usage but rather current. For all intensive purposes that is the resource we are guarding against anyway.

So I'd vote for a method on these oscf structs that calculates it's byte size representation in memory and we can take units from a semaphore against this value.

Reply:
> So I'd vote for a method on these oscf structs that calculates it's byte size representation in memory and we can take units from a semaphore against this value.

Isn't that just `sizeof(T)`?

Reply:
> I don't think we should measure future memory usage but rather current. For all intensive purposes that is the resource we are guarding against anyway.

I'd argue the start of back pressure is from the kafka client shard, which is "measured", if you will, by the size of the schemas.

Reply:
well this comment was really just that the wording of the configuration option needed to be cleaned up. not sure we need to hold up the entire PR because of the resource accounting for backpressure, but if it is unresolved it definitely needs to go onto the backlog as something to resolve.

Reply:
After discussing offline , we will punt this to a separate PR